LogBack 수정, MyBatis 연동, 프로필 환경 설정 yml로 변경, 단위 테스트 환경 구축

jieun-dev1 opened the pull request:
### **개요**

지난 1차 PR 에서 피드백 주신 것 위주로 개선하였습니다. 

### **이슈 링크**

#2 #3  #4  #8 

### **작업 사항 (변경 사항)**

**logBack 설정 수정/개선**

logback-dev.xml, logback-local.xml
-현재는 logger 별 로그 레벨 별도 설정이 필요치 않아서, root level 만 사용하는 것으로 변경. 
개발 환경까지는 DEBUG 로 자세한 로그 파악이 필요하다고 판단하여 local, dev 의 로그 레벨을 INFO -> DEBUG 변경.

logback-prod.xml
-폴더 위치를 LOG_PATH, LOG_FILE_NAME PropertyName 으로 지정.
-운영 환경에는 File appender 를 사용하기 위해서 appender ref="FILE"추가

file-appender.xml 
-file-appender에 하루 단위로 로그가 롤링되도록 설정. 확인 완료.

**MyBatis 연동** 

MarketGolaApplication (basePackageScan 관련)
-기존 구조는 controller 만 스캔하고 service, repository는 수동 빈 등록 하였음. 효율적이지 않다고 판단하여 전체 컴포넌트 스캔으로 대체.

MyBatisConfig
@bean 이 수동으로 중복등록 되어서 제거함.

**프로필 환경 설정**

-계층적 구조의 장점을 활용하기 위해 application.properties 를 yml 파일로 변경
-불필요한 로그 리포트 생성을 방지하기 위해서 logging.level에 autoconfigure=ERROR 설정 추가 

**단위 테스트**

- test/resources 하위 application.properties 에 단위 테스트전용 DB 연결.
-(리팩토링) 기존 코드인 .getId()는 .isPresent()와 함께 사용해야 했음.
이미 생성된 기본 객체를 찾아 제공하는 경우이므로, Optional.orElse(null)를 통해 코드를 간결하게 변경.

**[기타] 관례를 준수하지 않은 패키지/파일을 수정**

-console-appender.xml과 file-appender.xml 앞의 logback 을 제거
-package 는 소문자로 한 단어로 표현되는 관례를 반영하여, marketGola -> marketgola, User -> user 로 수정.
-RestAPI 규약을 준수하여, url 의  resources를 복수로 변경 

### 테스트 내역 (o/x)

**logback 설정** 

프로필 변경에 따라 세팅 대로 로그가 기록됨을 확인했습니다. 
지난 PR 에서는 운영 환경(PROD)에서 로그 파일에 날짜 형태 저장이 되지 않았는데, 현재는 날짜 형태 잘 반영 되었습니다. 
하루 단위로 파일 롤링 됨을 확인했습니다.  

**단위테스트 환경 구축**

UserRepositoryTest (유저 가입, 유저 단건 조회) 를 작성해두었고,  DB 접속과 데이터 저장 확인 완료, Test 통과 확인 하였습니다. 
@Transactional 을 달아서 테스트 환경에서는 데이터 커밋이 안되도록 설정했습니다.
조금 더 확실히 확인하고 싶어서, @Commit 을 추가해서 실제로도 잘 저장 됨을 잘 확인했습니다.

### **리뷰 받고 싶은 내용**

Optional을 활용한 단위 테스트 코드의 Then (검증)을 수정 하였습니다. Optional과 예외처리를 잘 알지 못해서
좋은 코드인지 리뷰 받고 싶습니다.

질문:  
Optional 을 쓰는 이유가 null을 피하기 위함이라고 알고 있는데, orElse(null)을 하게 되면 null을 참조하게 되는 것 아닌가요? 
테스트시에는 문제가 없는데, orElse(null)을 써도 괜찮을까요? 

참조한 자료들도 함께 공유드립니다. 

아래 블로그에서는 orElse(null)을 Optional처리의 좋은 사례라고 소개했고, 스택오버플로우에서도 이렇게 처리하라고 답을 준 것 같은데요. 
https://homoefficio.github.io/2019/10/03/Java-Optional-%EB%B0%94%EB%A5%B4%EA%B2%8C-%EC%93%B0%EA%B8%B0/
https://stackoverflow.com/questions/53652866/how-to-replace-return-optobj-ispresent-new-obj2optobj-null-in-a-functi 

해당 글의 댓글을 보면 제가 잘못 해석한 걸 수도 있겠지만, orElse(null)을 쓰는게 null을 참조하는 것 (Optional 쓰는 의미가 없음) 이나 다름 없다는 부분도 있는 것 같아서요.  
https://stackoverflow.com/questions/52047479/optional-ispresent-vs-orelsenull#comment91053600_52047479

기존 

```
 @Test
    void save() {
        //given
        User user = new User("JK", "jk@gmail1.com" );

        //when
        User savedUser = userRepository.save(user);

        //then
        User findUser = userRepository.findById(user.getId()).get();
        assertThat(findUser).isEqualTo(savedUser);
    }
```
1차 수정 

```
//        if (findUser.isPresent()) {
//            assertThat(findUser.get()).isEqualTo(savedUser);
//        }
```
수정한 이유
Optional.get()은 .isPresent()로 존재 여부를 먼저 확인한 후 써야 한다고 해서, 아래와 같이 수정하였더니, IDE 에서 can be replaced with single expression in functional style 경고를 주었습니다. 그래서 orElse(null)을 사용하였습니다. 

2차 수정 
```
//Then
 Optional<User> findUser = userRepository.findById(user.getId());
assertThat(findUser.orElse(null)).isEqualTo(savedUser);
}
```

jieun-dev1 committed:
초기구축 - 프로젝트 폴더 구성

jieun-dev1 committed:
LOG-BACK 설정파일 추가

jieun-dev1 committed:
초기구축 - 구동 환경을 local/dev/prod 로 나누기

jieun-dev1 committed:
초기구축 - mybatis 연동 위해 mysql과 mybatis 의존성 추가

jieun-dev1 committed:
초기구축 #4 - domain 중심으로 폴더 구조 변경.

jieun-dev1 committed:
rebase 작업 중 커밋

jieun-dev1 committed:
초기구축 #3 - application.properties 에 DB 설정 연결

jieun-dev1 committed:
초기구축 #3 - DataSource 설정

jieun-dev1 committed:
초기구축 #3 - MyBatis 설정 추가

(1) 언더 스코어 -> 카멜케이스 변경
(2) 로그 확인

jieun-dev1 committed:
chore: MyBatis 설정 추가 #3

오타 수정
profiles.active = local 설정 추가

jieun-dev1 committed:
chore: MyBatis 설정 추가 #3

test에 application.properties 추가

jieun-dev1 committed:
chore: MyBatisConfig 설정 수정 #3

jieun-dev1 committed:
chore: application.properties 설정 수정 #3

mybatis.type-aliases-package 탐색 추가

jieun-dev1 committed:
chore: logger 문제 해결 위해 exclude 추가 #3

slf4j 삭제

jieun-dev1 committed:
chore: MyBatisConfig import 추가

jieun-dev1 committed:
chore: 불필요한 import 삭제

jieun-dev1 committed:
feat: MyBatisUserRepository 구현체 추가

jieun-dev1 committed:
feat: User 클래스와 생성자 추가

jieun-dev1 committed:
feat: User Controller, Repository, Service 추가

jieun-dev1 committed:
feat: User Mapper 클래스와 UserMapper.xml

sql 쿼리 추가

jieun-dev1 committed:
feat: testDataInit 추가

jieun-dev1 committed:
초기구축 #3 MyBatis 연동 1차 완성

UserController의 model, TestDataInit, UserRepository 삭제
-View 가 없기 때문에 model을 삭제
-현재 코드에서는 초기 데이터 적재가 필요하지 않아서 TestDataInit을 삭제
-UserRepository의 구현체가 변경될 가능성이 없어서, 인터페이스 코드를 삭제

User, UserComtroller, UserRepository, UserService
- mybatis 연동을 테스트하기 위해 User 폴더 구조로  Controller, Service, Repository 구조 추가

UserDto
-데이터 전송 용도 객체 생성.

jieun-dev1 committed:
초기구축 #3 - Build.gradle 세팅

-logback-classic 을 최신 버전으로 업데이트
- mysql용 jdbc 드라이버, 마이바티스 추가.

jieun-dev1 committed:
초기구축 #4 - 구동 환경 나누기

application-local.properties
-로컬 환경 세팅. 로컬 전용 데이터베이스 marketGola 연결

application.properties (test 패키지), UserRepositoryTest
-단위 테스트 환경 세팅. 테스트 전용 데이터 베이스 marketgola_testcase 연결

application-dev.properties, application-prod.properties
-아직 초기 세팅 단계이므로 파일만 생성

jieun-dev1 committed:
초기구축 #2 - logback 설정

logback-file-appender.xml
-운영 환경에서는 파일로 로그를 남겼습니다.
-최상위 위치에 logs 폴더를 만들어 logback_log 이름으로 로그가 쌓이도록 함.

logback-dev/local/prod.xml
-dev, local, prod 환경 별로 logback.xml 설정을 나눔

logback-console-appender.xml
-중복을 줄이기 위해 환경 별로 겹치는 console-appender.xml을 별도로 분리.

jieun-dev1 committed:
초기구축 #4 - 구동환경 나누기

주석이 ?로 표시되어 수정하였습니다.

jieun-dev1 committed:
초기구축 #2 - logback 수정

logback-file-appender.xml
-최상위 위치에 logs 폴더를 만들어 logback.{현재 시간}.log 이름으로 로그가 쌓이도록 함

logback-dev/local/prod.xml
-dev, local, prod 환경 별로 logback.xml 설정을 나눔
-prod 에 file-appender.xml 을 추가함.

jieun-dev1 committed:
Merge pull request #7 from f-lab-edu/feature/db_setting

logback 설정, mybatis 연동, 프로필 별 구동 환경, 프로젝트 구조 구성

jieun-dev1 committed:
refactor(resources): 구동 환경 나누기 (#4)

-계층적 구조의 장점을 활용하기 위해 application.properties 를 yml 파일로 변경
-불필요한 로그 리포트 생성을 방지하기 위해서 logging.level에 autoconfigure=ERROR 설정 추가

jieun-dev1 committed:
chore(logging): 관례를 준수하지 않은 패키지/파일을 수정

-console-appender.xml과 file-appender.xml 앞의 logback 을 제거
-package 는 소문자로 한 단어로 표현되는 관례를 반영하여, marketGola -> marketgola, User -> user 로 수정.

jieun-dev1 committed:
refactor (test.user.repository: Optional 코드 수정 (#8)

- 기존 코드인 .getId()는 .isPresent()와 함께 사용해야 했음.
이미 생성된 기본 객체를 찾아 제공하는 경우이므로, Optional.orElse(null)를 통해 코드를 간결하게 변경.

jieun-dev1 committed:
chore (resource.logging): logback dev 환경의 로그 레벨 수정 (#2)

logback-dev.xml, logback-local.xml
-현재는 logger 별 로그 레벨 별도 설정이 필요치 않아서, root level 만 사용. 개발 환경까지는 DEBUG 로 자세한 로그 파악이 필요하다고 판단하여 INFO -> DEBUG 변경.

logback-prod.xml
-폴더 위치를 LOG_PATH, LOG_FILE_NAME PropertyName 으로 지정.
-운영 환경에는 File appender 를 사용하기 위해서 appender ref="FILE"추가

jieun-dev1 committed:
feat (resource.logging): logback dev 환경의 로그 레벨 수정 (#2)

-file-appender에 하루 단위로 로그가 롤링되도록 설정. 확인 완료.

jieun-dev1 committed:
chore (user): restAPI규약에 맞게 url resourec는 복수로, 에러메시지 영어로 수정

-user -> users
-한글 깨짐을 우려하여 에러메시지는 영어로 수정

jieun-dev1 committed:
refactor (user): @bean 제거 (#3)

MarketGolaApplication
-기존 구조는 controller 만 스캔하고 service, repository는 수동 빈 등록 하였음. 효율적이지 않다고 판단하여 전체 컴포넌트 스캔으로 대체함.

MyBatisConfig
-@bean 이 수동으로 중복등록 되어서 제거함.

derek-Flab commented on build.gradle:
```diff
@@ -0,0 +1,52 @@
+	runtimeOnly 'mysql:mysql-connector-java'
+
+	//log 세팅
+//	implementation 'ch.qos.logback:logback-classic:1.1.5'
```
주석 처리되어 있는 것은 특별한 경우를 제외하면 PR전에 삭제해주시면 좋을 것 같아요!

derek-Flab commented on build.gradle:
```diff
@@ -0,0 +1,52 @@
+
+	//log 세팅
+//	implementation 'ch.qos.logback:logback-classic:1.1.5'
+	implementation 'ch.qos.logback:logback-classic:1.2.3'
```
gradle에서 버전을 상수처럼 관리하도록 해서 사용합니다. 디펜던시 버전에 대한 관리를 쉽기 하기 위해서죠. 
gradle variable ext 이런식으로 검색해보면 사용법이 나올 겁니다!

derek-Flab commented on settings.gradle:
```diff
@@ -0,0 +1 @@
+rootProject.name = 'demo'
```
application네이밍 포함 demo가 아니라 marketgola, marketGolaApplication이렇게 되는게 좋겠습니다.

derek-Flab commented on src/main/java/com/marketGola/user/config/MyBatisConfig.java:
```diff
@@ -0,0 +1,27 @@
+public class MyBatisConfig {
+    private final UserMapper userMapper;
+
+//    @Bean
```
사용하지 않는 코드는 제거요!

derek-Flab commented on src/main/java/com/marketGola/MarketgolaApplication.java:
```diff
@@ -0,0 +1,14 @@
+import org.springframework.boot.autoconfigure.SpringBootApplication;
+import org.springframework.context.annotation.Import;
+
+@Import(MyBatisConfig.class)
```
Import를 꼭 해주어야 하는 것일까요? 
`@Configuration`으로 선언되어 있어도 못 읽어 올까요? 확인 테스트 한번 진행하면 좋을 것 같습니다.

jieun-dev1 replied:
config 파일 자체가 없어도 되더라고요! 확인 테스트 하고 삭제했습니다.

derek-Flab commented on src/main/java/com/marketGola/user/controller/UserController.java:
```diff
@@ -0,0 +1,32 @@
+
+    @GetMapping("/{userId}")
+    public User user(@PathVariable long userId) {
+        return userService.findById(userId).orElseThrow(() -> new NullPointerException("USER NOT FOUND"));
```
userService에서 조회하고 없으면 NullPointerException을 던지는 로직인데요. 

1. 보통 NPE를 던지는 경우는 없고, UserNotFoundException처럼 커스텀 예외를 정의해서 가독성 및 사용성을 높이고요! (임시 코드라 그런 것이겠죠?)
2. 없으면 예외를 던지는 로직이라면 service 레이어에 위치하는게 맞습니다. controller에서는 controller의 역할만 하는게 맞습니다. 프레젠테이션 역할에만 집중하도록 수정해주세요. request를 받아서 체크 및 정제하고, service레이어로 넘겨주는 역할만 합시다.

derek-Flab commented on src/main/java/com/marketGola/user/domain/User.java:
```diff
@@ -0,0 +1,26 @@
+    private String name;
+    private String email;
+
+    public User() {
```
lombok을 활용해서 빈생성자의 코드를 생략하면 좋을 것 같네요. 그리고 `@Data`가 있으면 빈생성자는 필요하지 않을텐데요?

jieun-dev1 replied:
네! @Data 는 생성자 RequiredArgsConstructor 를 포함하고 있어서, @NoArgsConstructor는 따로 추가하겠습니다.

derek-Flab commented on src/main/java/com/marketGola/user/service/UserService.java:
```diff
@@ -0,0 +1,26 @@
+        userRepository.save(user);
+    }
+
+    public Optional<User> findById(Long id) {
```
user가 없는 경우 null로 리턴할지, user가 없는 경우 예외를 던질지 메서드를 분리해봅시다. 
findByIdOrElseThrow(Long id)
findByIdOrElseNull(Long id)

이런식으로요~

derek-Flab commented on src/main/java/com/marketGola/user/service/UserService.java:
```diff
@@ -0,0 +1,26 @@
+public class UserService {
+    private final MyBatisUserRepository userRepository;
+
+    public void save(UserDto userDto) {
```
유저의 생성과 수정을 동시에 처리하실 건가요? 
등록과 수정을 분리하고 등록에 필요한 정보와 수정에 필요한 정보를 달리해서 받는게 좋을 것 같습니다.

jieun-dev1 replied:
멘토님 이 부분 다시 읽어보니 이해가 잘 가지 않아서요! 아직 수정 코드는 작성하지 않았고, 해당 UserService는 
(1) UserDto (user 가입을 위해서 요청해서 받아온 정보들) 을 받아서 save 
(2) User 가 있는지 조회하는 findById 입니다 (단건 조회) 
user.setUsername 을 하는 방법도 있을 것 같은데, 요청 정보이니 DTO 에서 정보를 꺼내서 User 를 생성해도 된다고 판단했습니다.

derek-Flab replied:
네 요건 좀 더 구현해보시고 정리해도 좋을 것 같습니다. 

save라는 네이밍이 생성 + 수정을 동시에 하나? 라는 착각을 일으켰던것 같습니다. 

save는 저장이라는 의미로 해석이 되어서요~

생성은 create, 수정은 update이렇게 가도 좋을 것 같습니다!

derek-Flab commented on src/main/resources/application-local.yml:
```diff
@@ -0,0 +1,22 @@
+logging:
```
이 설정은 반드시 필요한 것인가요? 
logback에서 적절히 설정해도 되지 않을까요? 
(가능하면 같은 영역의 설정은 한곳에서 관리하는게 좋을 것 같습니다. 유지보수 측면을 위해서요.)

jieun-dev1 replied:
mybatis:trace 설정은 Logback으로 옮겼습니다. org.springframework.boot.autoconfigure:ERROR 설정을 해줘야, Conditions evaluations report 라는 딱히 필요치 않은 리포트가 제거되더라고요. (참고: https://stackoverflow.com/questions/50273570/how-can-i-exclude-the-conditions-evaluation-report-from-the-console-of-a-spring) 이 설정은 application.yml 에서만 가능한 것 같아서 그대로 두었습니다.

derek-Flab commented on src/test/java/com/marketGola/user/repository/UserRepositoryTest.java:
```diff
@@ -0,0 +1,37 @@
+        Optional<User> findUser = userRepository.findById(user.getId());
+            assertThat(findUser.orElse(null)).isEqualTo(savedUser);
+        }
+//        if (findUser.isPresent()) {
```
사용하지 않는 코드는 제거 후 PR 부탁드려요~

derek-Flab commented on src/test/java/com/marketGola/user/repository/UserRepositoryTest.java:
```diff
@@ -0,0 +1,37 @@
+import static org.assertj.core.api.Assertions.assertThat;
+
+@Transactional
+@SpringBootTest
```
명시적으로 profiles를 명시하는 어노테이션을 추가했으면 좋겠습니다. !
`@ActiveProfiles("test")`

derek-Flab commented on src/test/resources/application.yml:
```diff
@@ -0,0 +1,20 @@
+        url:
+        username:
+    profiles:
+        active: test
```
이 프로파일 설정은 정상적으로 수행된다고 하더라도 
test resource패키지에 있어서 동작한 것이지 정상적인 설정은 아닐 겁니다. 
https://spring.io/blog/2020/08/14/config-file-processing-in-spring-boot-2-4

derek-Flab commented on src/test/java/com/marketGola/user/repository/UserRepositoryTest.java:
```diff
@@ -0,0 +1,37 @@
+        //then
+        Optional<User> findUser = userRepository.findById(user.getId());
+            assertThat(findUser.orElse(null)).isEqualTo(savedUser);
+        }
```
인텔리제이 포맷터 적용 안된거 같은데 확인해주세요~

derek-Flab commented on src/test/java/com/marketGola/user/repository/UserRepositoryTest.java:
```diff
@@ -0,0 +1,37 @@
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+@Transactional
```
메서드 단위로 transactional을 걸고, 단위 테스트로 격리하는게 좋겠습니다. 
각 단위 테스트에 저장된 값이 다른 테스트에 영향을 미칠 수 있습니다. 이렇게 되면 적절한 단위 테스트가 아니게 되겠죠?

derek-Flab commented:
리뷰 받고 싶은 내용에 대해서 답변 드리자면요. 
orElse(null)을 단위 테스트에서 처리해서 넣는건 엄밀히 말하면 단위 테스트가 아니고, 
서비스에서 이를 수행하는게 맞고요. 결과 값에 대해서만 처리하는게 맞습니다. 

우선 첫번째로 Assertions를 junit5에서 제공해주는 junit jupiter에서 가져와야 하고요~(아마 강의를 들으시고 하셔서 혼란스러우셨을 겁니다)

그래서 현재의 경우 
`assertThat(findUser.orElse(null)).isEqualTo(savedUser);`이 아니라


junit5 jupiter에서 제공하는 
`Assertions.assertFalse(findUser.isEmpty());`가 맞을 것 같습니다. 
Assertions을 라이브러리에서 가지고 올때 jupiter쪽에서 가지고 오도록 해주시고요. 
그리고 이 jupiter에서 가져오는게 무엇인지에 대해서 정리 한번 하면 좋을 듯 하네요~!

jieun-dev1 committed:
chore (resources): yml 수정 (#4)

-test/resources 폴더의 application.yml 제거. main/resources 의 application-test.yml 추가
-mybatis:trace 설정을 logback 에 추가
-marketGola -> marketgola 로 패키지 명 수정 반영

jieun-dev1 committed:
chore : build.gradle 버전 상수화

버전관리의 용이함 위해서 ext 로 버전 관리

jieun-dev1 committed:
chore : logback 파일 포맷 수정 (#2)

커밋된 파일 공통 사항
-구글 포맷터 적용으로 인한 간격 수정

logback-dev, prod.xml
-mybatis trace 설정 추가.

jieun-dev1 committed:
chore : 불필요한 DB 관련 파일 삭제 (#3)

MyBatisConfig 삭제
-Config 파일 없이도 정상 구동 되어 삭제함.

MyBatisUserRepository 삭제 후 userMapper 로 대체

jieun-dev1 committed:
chore: rootProject name 수정

jieun-dev1 committed:
refactor: User 관련 초기 구현 추가

공통사항
-구글 포맷터 적용

User
-생성자 롬복으로 대체

UserService
-findByIdOrElseThrow(예외 처리)와 findByIdOrElseNull으로 메서드 분리
-save 보다 명확한 신규가입을 나타내는 ㄴregister 로 변경.
-validateDuplicateMember(회원 이름 중복 체크 추가)

UserMapper
-findByName 추가

UserController
-dupCheck 중복 체크 기능 추가
-ResponseEntity로 반환 타입 변경

jieun-dev1 committed:
refactor: User 관련 초기 구현 추가

User-mapper
-findByName 추가.
-Package 이름 수정

jieun-dev1 committed:
refactor: 테스트 리팩토링 (#6)

-package 이름 demo -> marketgola 로 수정

-스프링 프레임워크에 의존하지 않도록 Mockito를 사용한 UserServiceTest 추가. 서비스에서 mapper method 테스트도 함께 하기 때문에, UserRepositoryTest를 삭제함.

-스프링 프레임워크 통합 테스트 UserIntegrationTest  추가

derek-Flab commented on src/main/java/com/marketGola/user/controller/UserController.java:
```diff
@@ -0,0 +1,51 @@
+
+    @GetMapping("/{userId}")
+    public Optional<User> findUser(@PathVariable Long userId) {
+        return userService.findByIdOrElseNull(userId);
```
이 API의 response도 다른 API와 동일하게 ResponseEntity로 감싸서 응답해주는게 일관성 있을 것 같습니다.

derek-Flab commented on src/main/java/com/marketGola/user/service/UserService.java:
```diff
@@ -0,0 +1,36 @@
+    }
+
+
+    public User findByIdOrElseThrow(Long id) {
```
`findByIdOrElseThrow{익셉션이름}(Long id)`로 네이밍하는게 어떨까요? 
명시적으로 어떤 exception을 던지는지요. 

-> findByIdOrElseThrowIllegalState(Long id)

derek-Flab commented on src/main/java/com/marketGola/util/HttpResponses.java:
```diff
@@ -0,0 +1,12 @@
+package com.marketgola.util;
```
이게 만약 있다면 DTO패키지에 있는게 더 낫지 않을까요?

jieun-dev1 replied:
현재는 request 와 response 가 있는데요 (패키지에 파일이 없어서 푸시가 되지 않아서 git 에 보이지 않았네요!) DTO 를 request와 response 로 나눌 수 있을 것 같아서 dto 패키지는 만들지 않았습니다. 이 경우에,  response 패키지에 있는 게 더 적절하지 않을까요? 혹은 dto 패키지도 별도로 만들 필요가 있을까요?

derek-Flab replied:
네네 
1. request, response패키지를 나누는 것 찬성이고요. 
2. dto패키지에 담아도 괜찮을 것 같구요.

저는 1번입니다. 
어찌됐든 util패키지에 있는 것 은 아닌 것 같습니다!

derek-Flab commented on src/main/java/com/marketGola/util/HttpResponses.java:
```diff
@@ -0,0 +1,12 @@
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+
+public class HttpResponses {
```
자주 사용하는 경우라 판단되어서 이렇게 상수 클래스를 선언하여 쓰도록 만드신걸까요?
ResponseEntity(HttpStatus.OK) 만으로도 읽히는데 문제가 없을 것 같아서 따로 빼지 않아도 될 것 같다는 생각이 듭니다. 
그리고 보통 ResponseEntity에 body를 채워넣는 경우가 많은데, 아래와 같은 경우는 body없는 responseEntity를 모아놓는 용도로만 활용되겠죠? 음 어떤 고민이 있었고, 어떤 아이디어인지 공감가지만! 
이 부분은 좀 더 확장될 만한 포인트가 있을지 한번 더 생각해보면 좋을 것 같습니다. 
단순히 HTTP status를 경우의 수대로 정의한 ResponseEntity로서는 크게 필요성을 공감하기 힘든 것 같습니다.

derek-Flab replied:
추가로 ResponseEntity는 제네릭 객체로 타입을 명시해주면 좋을 것 같습니다. 없다면  `<Void>`이렇게라도요.

derek-Flab commented on src/test/java/com/marketGola/user/UserIntegrationTest.java:
```diff
@@ -0,0 +1,67 @@
+                () -> userService.validateDuplicateMember(user2.getName()));
+        assertThat(e.getMessage()).isEqualTo("name ALREADY TAKEN");
+
+        userService.validateDuplicateMember(user3.getName());
```
테스트 코드는 단순하고 명료했으면 좋겠습니다. 테스트 시나리오는 다음과 같죠? 

user1이 이미 저장되어있고 -> 이거와 같은 이름을 쓰면 throw가 수행되는거죠? -> 그리고, user3을 넣으면 정상 동작하는 것이고요. 
이거는 논리적으로는 분리된 테스트이어야 할 것 같습니다. 

TEST1) 
user1이 저장됩니다.(given) -> user2를 저장합니다. -> duplicate throw 가 발생합니다. 

TEST2) 
user1이 저장됩니다.(given) -> user3을 저장합니다. -> 정상 수행됩니다. 

위의 테스트 2개가 하나로 해결되고 있는 것 같아요.

derek-Flab replied:
다른 테스트들도 이런게 있다면 수정해보면 좋을 것 같습니다!

jieun-dev1 replied:
네 수정해보겠습니다! 
통합 테스트에 대해 질문 있습니다. 
해당 테스트를 작성할 때, 제가 생각한 통합 테스트는 서비스 테스트와 같은 케이스를 테스트하지만, 스프링 프레임워크에 의존한다/db연결 정도의 차이로만 생각했었습니다 (김영한님 스프링 입문 강의 참고)  실제로도 이렇게 작성하나요? 추가로 고려해야 할 게 있을까요?

forest-lim-dev replied:
실제로도 이렇게 하는가는 정확히 규칙이 있는건가? 라는 질문이라면 없습니다. 
어느 팀은 단위 테스트를 하기위해 모든 bean을 목킹한다고 하는 팀도 있고, 그렇게 하면 코드가 수정되었을 때 영향범위가 너무 커서 DB를 연동하는 지은님이 말씀하신 service intergration 테스트를 하는 곳도 있습니다. 

저희 팀은 service레이어에 대해서 외부 연동 부분(API)만 목킹하고 H2 DB에 직접 쓰는 것을 단위 테스트로 하고 있습니다. 
integration + unit 테스트라는 개념의 중간지점에 있다고 할 수 있을 것 같습니다.

derek-Flab commented on src/test/java/com/marketGola/user/service/UserServiceTest.java:
```diff
@@ -0,0 +1,86 @@
+
+    @Test
+    @DisplayName("회원 가입 성공")
+    void register() throws Exception {
```
// given // when // then주석을 모두 표시하는 형태로 연습을 계속 해보는게 어떨까요? ㅎㅎ
(필수는 아님)
저는 보통 이렇게 모든 테스트에 주석을 선언해놓고 테스트 코드를 작성합니다. (일종의 일관성 유지라고 할까요?)

derek-Flab reviewed (commented):
공통 예외 처리 부분은 추가적으로 작업해주고 계신거죠1?ㅎㅎ화이팅입니다!!

derek-Flab commented:
그리고 package명을 수정할 때는? 인텔리제이에서 제공하는 rename기능을 활용해주세요. 그래야 프로젝트 전반적으로 일괄 수정 됩니다. 
현재 package명은 com.marketGola인데 각 클래스의 package명은 com.marketgola입니다.

jieun-dev1 committed:
chore(util): Response에 제네릭 타입 추가

-리턴 타입 없는 경우 void 추가.

jieun-dev1 committed:
refactor(user, resources):서비스 로직 수정

User 관련 기능 전체
-name을 loginId로 수정

UserMapper
-findById 삭제 (회원 조회, 회원 중복 체크 모두 loginId로 하기 때문)

UserController, UserService, user-mapper.xml
-로그인 아이디 중복 확인 후 가입으로 변경
-가입에서 예외처리를 하는 것이 맞다고 판단하여,
아이디 중복 확인 로직을 예외처리에서 Boolean 응답으로 변경
-findUser: Long 값이 아닌 고유값인 loginId로 회원 검색하도록 변경
-findByOrElseNull은 validateDuplicateNumber 와 중복되어 삭제

jieun-dev1 committed:
refactor(test):테스트 코드 리팩토링 (#8)

UserServiceTest
-중복 유무와 회원 가입 여부가 섞여있었던 코드를 케이스 별로 나눔
(중복 유저 유무/가입 성공 유무)

UserIntegrationTest
-한 테스트에 두 개 이상의 케이스가 있었던 것을 분리/ 하나의 케이스만 포함하는 것으로 수정

derek-Flab commented on build.gradle:
```diff
@@ -0,0 +1,44 @@
+
+dependencies {
+    //MyBatis
+    implementation "org.mybatis.spring.boot:mybatis-spring-boot-starter:${mybatisSpringBootStarterVersion}"
```
일관성 있게 `"` -> `'`으로 가시죠~

derek-Flab commented on src/main/java/com/marketGola/user/service/UserService.java:
```diff
@@ -0,0 +1,31 @@
+        userMapper.save(user);
+    }
+
+    public Boolean validateDuplicateMember(String loginId) {
```
👍

derek-Flab commented on src/test/java/com/marketGola/user/UserIntegrationTest.java:
```diff
@@ -0,0 +1,69 @@
+class UserIntegrationTest {
+
+    @Autowired
+    UserService userService;
```
빈에 습관적으로 private을 붙여주는게 좋을 것 같습니다!

jieun-dev1 replied:
테스트코드여서 해당 클래스 내에서만 작동할 예정이어서 private 을 쓰지 않았는데요. 그럼에도 private 을 붙여주는 걸 권장해주신 이유는, 외부에서 임의로 값을 변경할 위험을 방지하기 위해서겠죠?

jieun-dev1 replied:
테스트코드여서 해당 클래스 내에서만 작동할 예정이어서 private 을 쓰지 않았는데요. 그럼에도 private 을 붙여주는 걸 권장해주신 이유는, 외부에서 임의로 값을 변경할 위험을 방지하기 위해서겠죠?

sgo8308 replied:
해당 클래스 내에서만 작동할 예정이기 때문에 명확하게 private을 붙여주라는 의미가 아닐까요?
처음 보는 사람 입장에서는 private이 안 붙어 있으면 다른 곳에서도 쓰이나?라고 생각할 여지가 있을 수도 있다고 생각합니다.

forest-lim-dev replied:
네네 지우님이 말씀하신 의도로 리뷰 드린 것 맞습니다.
명시적으로 클래스 내에서만 사용할 private 할 목적이면 모두 붙여달라는 의미였습니다. 
테스트여서 문제없다거나, 이건 구조적으로 문제 없다거나 하는 것보다는 명시적으로 선언하는게 좋겠다는 의견으로 이해해주시면 될 것 같습니다.

derek-Flab commented on src/test/java/com/marketGola/user/service/UserServiceTest.java:
```diff
@@ -0,0 +1,95 @@
+import org.mockito.junit.jupiter.MockitoExtension;
+
+@ExtendWith(MockitoExtension.class)
+class UserServiceTest {
```
integration 테스트와 단위테스트를 분리하는 방법으로 구현하셨네요? 
1. 단위 테스트에서는 bean들을 모두 mocking하고 
2. 인테그레이션 테스트에서만 db를 연계한다.

이런 방식이네요?

jieun-dev1 replied:
네 맞습니다! 우선 통합 테스트는 어떤 식으로 짜야하는지 잘 감이 잡히지 않아서, 서비스 테스트와의 차이점을 (1) DB 연계 (2) 스프링 프레임워크에 의존 (@SpringBootTest, @Transactional 등) 으로만 두었습니다. 스프링 입문 김영한님 강의 24-25(서비스 테스트), 45-46(통합 테스트) 에서도 테스트 시나리오는 비슷하고 위의 차이만 있더라고요. 
정답은 없는 것 같아서, 다른 프로젝트도 참고해보았는데 안한 경우도 있고, 프로젝트 마다 조금씩 다른 것 같아서요. 실제 구현을 할 때는 어떻게 할지 아직 고민해보고 있습니다!  통합 테스트를 짤 때, 단위 테스트와는 어떤 차이가 있어야 할까요?

forest-lim-dev replied:
이 부분은 멘토링 때 한번 다시 이야기해보면 좋을 것 같네요!

jieun-dev1 committed:
chore: bean 을 private 생성자로 등록 (#8)

클래스 내에서만 사용하기 때문에, private 명시

jieun-dev1 committed:
chore: dependencies를 일관성 있게 큰 따옴표로 표기

jieun-dev1 committed:
chore: 패키지 이름을 marketGola에서 marketgola 로 수정
